spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,186 @@
+package org.apache.spark.sql.cassandra
+
+import java.io.IOException
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.ValidRDDType
+import com.datastax.spark.connector.writer.SqlRowWriter
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.catalyst.expressions.{Literal, AttributeReference, Expression, Attribute}
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+import scala.reflect.ClassTag
+
+
+case class CSPrunedFilteredScanRelation(table: String,
+                                   keyspace: String,
+                                   cluster: Option[String],
+                                   userSpecifiedSchema: Option[StructType],
+                                   sqlContext: SQLContext) extends PrunedFilteredScanRelationImpl
+
+case class CSBaseScanRelation(table: String,
+                               keyspace: String,
+                               cluster: Option[String],
+                               userSpecifiedSchema: Option[StructType],
+                               sqlContext: SQLContext) extends TableScanRelationImpl
+
+case class CSPrunedScanRelation(table: String,
+                               keyspace: String,
+                               cluster: Option[String],
+                               userSpecifiedSchema: Option[StructType],
+                               sqlContext: SQLContext) extends PrunedScanRelationImpl
+
+case class CSCatalystScanRelation(table: String,
+                                keyspace: String,
+                                cluster: Option[String],
+                                userSpecifiedSchema: Option[StructType],
+                                sqlContext: SQLContext) extends CatalystScanRelationImpl
+
+case object BaseScanType extends ScanType
+case object PrunedScanType extends ScanType
+case object PrunedFilteredScanType extends ScanType
+case object CatalystScanType extends ScanType
+
+trait ScanType
```

Comment:
I think that it would be better if `ScanType` had a method declaration `makeRelation` or something which would be implemented in particular case objects. Then, you would need only one case class, like `Relation` - then you could create particular types dynamically like `new Relation(...) with PrunedFilteredScanRelationImpl`